### ISSUE TITLE
How is the dynamic head more efficient than the vanilla head?

### ISSUE THREAD

**Issue by cmsflash** [WXinlong/SOLO#94]
I read your papers on SOLO and SOLOv2. One part that I didn't understand was that you said the dynamic version is more efficient than the vanilla version. How is efficiency measured here? If it's measured by computation (i.e. MACC/FLOP), then since you eventually generate the HxWxS^2 feature map, isn't the efficiency similar? If it's measure by parameters, than the dynamic version is indeed more efficient.

**Comment by ysyyork** [comment]
as far as i understand, the efficiency is measured by how much memory it consumes for sure and maybe computation (which I'm not sure):

![image](https://user-images.githubusercontent.com/6693605/92469572-9c591c80-f207-11ea-85fb-a2c337a63ce1.png)

As per the paper shows here, the two HxWxS branches become one branch of SxSxD and usually S << H or W and D << W or W. So the memory footprint for the v2 is much smaller than the decoupled one. As computation, since the v2 is using convolution to combine the two branch vs the decoupled one using elem wise multiplication, I'm not too sure which one consumes less but I think if you spend some time you can calculate it manually. Hope this helps. I'm not the author so this can be wrong.

**Comment by cmsflash** [comment]
Hi York, thank you so much for answering. If you compare these two parts, than the dynamic head should be more efficient than the decoupled head. However, my main concern is that what is displayed in the figure is not where the bulk of memory/computation is. Since you eventually generate the **mask**, it should have shape `H * W * S^2`, much larger than any of the intermediate outputs. Therefore, all three variants (vanilla, decoupled, dynamic) should have similar costs (because anything other than the **mask** incurs comparatively negligible costs).

**Comment by ysyyork** [comment]
> Hi York, thank you so much for answering. If you compare these two parts, than the dynamic head should be more efficient than the decoupled head. However, my main concern is that what is displayed in the figure is not where the bulk of memory/computation is. Since you eventually generate the **mask**, it should have shape `H * W * S^2`, much larger than any of the intermediate outputs. Therefore, all three variants (vanilla, decoupled, dynamic) should have similar costs (because anything other than the **mask** incurs comparatively negligible costs).

no, for  both modes you don't need to form an `H x W x S^2` while training and that's the beauty. why you say you finally need a shape of this? If you are given i, j and the two HxWxS heads in decouple, you are able to index a mask in shape of HxW and then compare to the gt. Similar for the dynamic head. There are actually some tricks you can do on the implementation that can help shrink the memory footprint a lot with tensorflow. I have done some work on that.

**Comment by cmsflash** [comment]
How do you supervise they masks without generating them? Do we only generate masks that are matched to ground truths and leave other masks?

**Comment by ysyyork** [comment]
U can still generate the mask but just not in the shape if HxWxS^2，both
solutions are telling to do that. For the gt u passed into the model, u can
only pass in valid mask instead of a HxWxS^2 tensor. This will save u a lot
if memory. Prediction wise u don’t need to generate that big tensor either

**Comment by cmsflash** [comment]
I see. Thank you for your explanation! It's very helpful!
